G-Modules missing methods (part 1)

mjrodgers opened the pull request:
This is split off from #3266 and should be the "easy" part, this will make way for the rest of the work to be done addressing #3224 .

This should be good to merge

fieker committed:
experimental subfields now working (better)

fieker committed:
refine the is_disjoint

fieker committed:
export the lattice in some way

fieker committed:
improve linear disjoint and strategy if many factors

mjrodgers committed:
formatter

fieker committed:
fixes for "THE" book

fieker committed:
minor

mjrodgers committed:
Write `extension_of_scalars`, add `absolute_frobenius`

mjrodgers committed:
Fix `gmodule_over` for finite fields

mjrodgers committed:
add test for `extension_of_scalars`

thofma commented:
What does `@which x == y` say, when `x` and `y` are GModules?

mjrodgers commented:
> What does `@which x == y` say, when `x` and `y` are GModules?

I get
```
julia> @which (ME1 == ME2)
==(x, y)
     @ Base Base.jl:165
```
So I guess that means we don't have an implemented method to determine equality?

thofma commented:
This is the generic `==` fallback, which just calls `===`. You need to implement your own `==` for GModules.

fingolfin commented:
As a reminder, you implement `==` then you should also implement `hash`

fingolfin commented on experimental/GModule/test/runtests.jl:
```diff
@@ -44,6 +44,29 @@ end
+          matrix(F.([0 0 1; 1 0 0; 0 1 0])),
+          matrix(F.([2 0 0; 0 1 0; 0 0 2])),
+          matrix(F.([2 0 0; 0 2 0; 0 0 1])),
+          matrix(F.([1 0 0; 0 1 0; 0 0 1]))
```
This seems weird, shouldn't it be this (and then similar a few lines later)? I mean both will work, but the current code allocates more.
```suggestion
          matrix(F, [0 0 1; 1 0 0; 0 1 0]),
          matrix(F, [2 0 0; 0 1 0; 0 0 2]),
          matrix(F, [2 0 0; 0 2 0; 0 0 1]),
          matrix(F, [1 0 0; 0 1 0; 0 0 1])
```

fingolfin commented on experimental/GModule/test/runtests.jl:
```diff
@@ -44,6 +44,29 @@ end
+      ]
+  mE = free_module(E, 3)
+
+  @test Oscar.GModuleFromGap.extension_of_scalars(M, phi) == GModule(mE, G, [hom(mE, mE, a) for a in LE])
```
Shouldn't we export `extension_of_scalars`?

mjrodgers committed:
Update experimental/GModule/test/runtests.jl

Co-authored-by: Max Horn <max@quendi.de>

mjrodgers committed:
export `extension_of_scalars`

lgoettgens commented on experimental/GModule/GModule.jl:
```diff
@@ -1563,6 +1572,7 @@ end #module GModuleFromGap
 
 using .GModuleFromGap
 
+export extension_of_scalars
```
this needs to be added to line 1520 as well

mjrodgers committed:
export again

fieker committed:
invar._classes

mjrodgers committed:
fix formatting

mjrodgers committed:
remove methods that are already merged to `AbstractAlgebra.jl`

mjrodgers commented:
I've implemented `==` and a hash function for `GModule`, and incorporated some work by @fieker to support functionality needed for the book. There is some work that needs to be done on writing better tests, and in particular implementing a type for GModule homomorphisms, but we feel this can be merged for now and these other things can come as part of another pull request.

lgoettgens commented:
You need to bump the AA compat entry to 0.40.1 to make sure that the direct_sum stuff from AA is available.

fieker commented:
... and rebase?

lgoettgens commented:
`SubfieldLattice_Module.AnticNumberField is deprecated, use Nemo.AbsSimpleNumField instead.`

mjrodgers commented:
> ... and rebase?

Weird, I did rebase, but it looks like it didn't work. I will redo it (and make the other changes)

mjrodgers committed:
Update deprecated types

mjrodgers committed:
Update AA Dependency

mjrodgers commented:
Okay, rebased (successfully), removed outdated types, and updated the `AbstractAlgebra` dependency

lgoettgens commented on experimental/GaloisGrp/src/Subfields.jl:
```diff
@@ -440,4 +479,13 @@ function _subfields(K::AbsSimpleNumField; pStart = 2*degree(K)+1, prime = 0)
   end
 end
 
+function subfield_lattice(K::AbsSimpleNumFiels)
```
```suggestion
function subfield_lattice(K::AbsSimpleNumField)
```

lgoettgens committed:
Update experimental/GaloisGrp/src/Subfields.jl

mjrodgers committed:
Partial fix for failing supfield test

mjrodgers commented:
I'm trying to figure out what's going on with a weird error trying to compute the intersection of two subfield lattice elements, hopefully this will be something minor.

mjrodgers committed:
More partial fix for failing subfield test section

fingolfin commented on experimental/GModule/Brueckner.jl:
```diff
@@ -458,6 +474,11 @@ function lift(C::GModule, mp::Map)
 
+function solvable_quotient(G::Oscar.GAPGroup)
+  q = pc_group(symmetric_group(1))
+  mp = hom(G, q, [one(q) for g = gens(G)])
```
My personal preference
```suggestion
  mp = hom(G, q, [one(q) for g in gens(G)])
```

fingolfin commented on experimental/GModule/Brueckner.jl:
```diff
@@ -176,15 +179,20 @@ the p is in the set.
-  I = irreducible_modules(ZZ, Q) 
+  if order(Q) == 1
+    F = free_module(ZZ, 1)
+    I = [gmodule(F, Q, [hom(F, F, [F[1]]) for x = gens(Q)])]
```
```suggestion
    I = [gmodule(F, Q, [hom(F, F, [F[1]]) for x in gens(Q)])]
```

fingolfin commented on experimental/GModule/Brueckner.jl:
```diff
@@ -442,11 +451,18 @@ function lift(C::GModule, mp::Map)
 
-    h = hom(G, GG, gens(G), l)
+    h = try
+          hom(G, GG, gens(G), l)
```
I think the generators for the domain can be omitted and then default to `gens(G)`. So does this also work?
```suggestion
          hom(G, GG, l)
```

fingolfin commented on experimental/GModule/Brueckner.jl:
```diff
@@ -458,6 +474,11 @@ function lift(C::GModule, mp::Map)
 end
 
+function solvable_quotient(G::Oscar.GAPGroup)
+  q = pc_group(symmetric_group(1))
```
If you want a trivial pc group, perhaps this is a simpler way
```suggestion
  q = cyclic_group(1)
```

mjrodgers replied:
These seem to give subtly different groups in some way (`==` returns `false`, even though they are both Pc groups of order 1....), I'll check that this change doesn't cause problems.

fingolfin commented on experimental/GModule/Cohomology.jl:
```diff
@@ -170,7 +184,8 @@ end
   if !isdefined(C, :F)
-    if order(Group(C)) == 1
+    @show Group(C)
+    if (!isa(Group(C), FPGroup)) && order(Group(C)) == 1
```
meta-remark: does `group(C)` also work (with lowercase `g`)? Then we should use that.

Also, use `is_trivial(group(C))` ?

fingolfin commented on experimental/GModule/GModule.jl:
```diff
@@ -207,8 +212,71 @@ where `phi` is an embedding of `S` into `R`.
+    new  = false
+    lres = length(res)
+    for X = res[sres:end]
+      for p = lp
```
I really don't like the `=` in iteration, would you mind changing it to
```suggestion
    for X in res[sres:end]
      for p in lp
```

fingolfin commented on experimental/GModule/GModule.jl:
```diff
@@ -207,8 +212,71 @@ where `phi` is an embedding of `S` into `R`.
+        end
+      end
+    end
+    lres = length(res)
```
Redundant?
```suggestion
```
Also, with this gone, you could also get rid of variable `new` and instead do at the top:
```
  lres = 0
  while lres < length(res)
    lres = length(res)
```

mjrodgers replied:
This was Claus' code, I'm supposed to be looking it over to understand and see where improvements can be made; it seems like this change removing `new` leaves everything functioning the same, but I'm double checking everything.

fingolfin reviewed (commented):
A few minor comments

@mjrodgers what exactly is the error you are getting? Perhaps we can help you debug it.

mjrodgers committed:
Update experimental/GModule/Brueckner.jl

Co-authored-by: Max Horn <max@quendi.de>

mjrodgers committed:
Update experimental/GModule/Brueckner.jl

Co-authored-by: Max Horn <max@quendi.de>

mjrodgers committed:
Update experimental/GModule/GModule.jl

Co-authored-by: Max Horn <max@quendi.de>

mjrodgers commented:
> A few minor comments
> 
> @mjrodgers what exactly is the error you are getting? Perhaps we can help you debug it.

I'm getting the following error from the `intersect` method:
```julia
Zx, x = ZZ["x"]
k, a = number_field(swinnerton_dyer(3, x))
s = subfield_lattice(k)
length(s) == 14

intersect(s[3], s[4])
```
triggers
```
julia> intersect(s[3], s[4])
ERROR: BoundsError: attempt to access 2-element Vector{Vector{Int64}} at index [[3, 4]]
Stacktrace:
 [1] throw_boundserror(A::Vector{Vector{Int64}}, I::Tuple{Vector{Int64}})
   @ Base ./abstractarray.jl:737
 [2] checkbounds
   @ ./abstractarray.jl:702 [inlined]
 [3] _getindex
   @ ./multidimensional.jl:888 [inlined]
 [4] getindex
   @ ./abstractarray.jl:1291 [inlined]
 [5] intersect(A::Oscar.SubfieldLattice_Module.SubfieldLatticeElem{…}, B::Oscar.SubfieldLattice_Module.SubfieldLatticeElem{…})
   @ Oscar.SubfieldLattice_Module ~/.julia/dev/Oscar/experimental/GaloisGrp/src/Subfields.jl:141
 [6] top-level scope
   @ REPL[10]:1
```
Claus made some minor changes to the `intersect` method, but nothing that I can see that would obviously lead to this issue.

mjrodgers committed:
update `fp_group_with_isomorphism`

mjrodgers committed:
Update experimental/GModule/Brueckner.jl

Co-authored-by: Max Horn <max@quendi.de>

mjrodgers committed:
Update experimental/GModule/Brueckner.jl

Co-authored-by: Max Horn <max@quendi.de>

mjrodgers committed:
fix error in subfield lattice `intersect`

mjrodgers commented:
Okay, I think I ironed out the bug in `intersect` (it looked like there was a typo and the wrong array was being accessed)

fingolfin commented on experimental/GModule/Brueckner.jl:
```diff
@@ -11,13 +11,16 @@ function Hecke.roots(a::FinFieldElem, i::Int)
   return roots(x^i-a)
 end
 
+Oscar.matrix(phi::Generic.IdentityMap{<:AbstractAlgebra.FPModule}) = identity_matrix(base_ring(domain(phi)), dim(domain(phi)))
```
Should this be added to AbstractAlgebra (not now but down the road)?

fingolfin commented on experimental/GModule/Brueckner.jl:
```diff
@@ -210,9 +218,9 @@ function find_primes(mp::Map{<:Oscar.GAPGroup, PcGroup})
     TODO: this is not (yet) implemented this way
     =#
-    q = cokernel(b)[1]
+    @show q = cokernel(b)[1]
```
```suggestion
    q = cokernel(b)[1]
```

fingolfin commented on experimental/GModule/Brueckner.jl:
```diff
@@ -210,9 +218,9 @@ function find_primes(mp::Map{<:Oscar.GAPGroup, PcGroup})
+    @show q = cokernel(b)[1]
 #    q = quo(kernel(da)[1], image(db)[1])[1]
-    t = torsion_subgroup(q)[1]
+    @show t = torsion_subgroup(q)[1]
```
```suggestion
    t = torsion_subgroup(q)[1]
```

fingolfin commented on experimental/GModule/Brueckner.jl:
```diff
@@ -402,6 +410,7 @@ function lift(C::GModule, mp::Map)
   seen = Set{Tuple{elem_type(D), elem_type(codomain(mH2))}}()
   #TODO: the projection maps seem to be rather slow - in particular
   #      as they SHOULD be trivial...
+  global last_k = k
```
```suggestion
```

fingolfin commented on experimental/GModule/Brueckner.jl:
```diff
@@ -442,11 +451,18 @@ function lift(C::GModule, mp::Map)
       end
       return d
     end
+    @show [pro[i](epi) for i=1:ngens(G)]
```
```suggestion
```

fingolfin commented on experimental/GModule/Brueckner.jl:
```diff
@@ -442,11 +451,18 @@ function lift(C::GModule, mp::Map)
     l= [GMtoGG(reduce(gen(G, i)), pro[i](epi)) for i=1:ngens(G)]
-#    @show map(order, l), order(prod(l))
+    @show map(order, l), order(prod(l))
+    global last_x = l
```
```suggestion
#    @show map(order, l), order(prod(l))
```

fingolfin commented on experimental/GModule/Brueckner.jl:
```diff
@@ -442,11 +451,18 @@ function lift(C::GModule, mp::Map)
 #    @show map(order, gens(G)), order(prod(gens(G)))
 
-    h = hom(G, GG, gens(G), l)
+    h = try
+          hom(G, GG, l)
+        catch
+          @show :crash
+          continue
+        end
```
```suggestion
    h = hom(G, GG, gens(G), l)
```
or
```suggestion
    h = hom(G, GG, l)
```

fingolfin commented on experimental/GModule/Cohomology.jl:
```diff
@@ -170,19 +184,20 @@ end
   #TODO: better for PcGroup!!!
   if !isdefined(C, :F)
-    if order(Group(C)) == 1
+    @show Group(C)
```
```suggestion
```

fingolfin commented on experimental/GModule/GModule.jl:
```diff
@@ -207,8 +212,71 @@ where `phi` is an embedding of `S` into `R`.
+  C = Gap(M)
+  S = GAP.Globals.MTX.BasesMaximalSubmodules(C)
+  res = []
+  for s = S
```
```suggestion
  for s in S
```

fingolfin commented on experimental/GModule/GModule.jl:
```diff
@@ -207,8 +212,71 @@ where `phi` is an embedding of `S` into `R`.
+function maximal_submodule_bases(M::GModule{<:Oscar.GAPGroup, <:AbstractAlgebra.FPModule{<:FinFieldElem}})
+  C = Gap(M)
+  S = GAP.Globals.MTX.BasesMaximalSubmodules(C)
+  res = []
```
This leaves `res` as a `Vector{Any}`, perhaps we can improve that...
```suggestion
  res = dense_matrix_type(base_ring(M))[]
```

fingolfin commented on experimental/GModule/GModule.jl:
```diff
@@ -476,6 +544,10 @@ function Oscar.character_field(C::GModule{<:Any, <:AbstractAlgebra.FPModule{AbsS
 end
 
+function Oscar.character(C::GModule{<:Any, <:AbstractAlgebra.FPModule{QQAbElem{AbsSimpleNumFieldElem}}})
+  return Oscar.class_function(group(C), [x[2] for x = _character(C)])
```
```suggestion
  return Oscar.class_function(group(C), [x[2] for x in _character(C)])
```

fingolfin commented on experimental/GModule/GModule.jl:
```diff
@@ -556,21 +632,21 @@ function gmodule_minimal_field(C::GModule{<:Any, <:AbstractAlgebra.FPModule{<:Fi
 
 """
-    gmodule_over(Field)
+    gmodule_over(Field, GModule)
```
```suggestion
    gmodule_over(k::Field, C::GModule)
```

fingolfin commented on experimental/GModule/GModule.jl:
```diff
@@ -898,11 +978,17 @@ A needs to have norm 1 wrt. to s, so
-function hilbert90_cyclic(A::MatElem{<:FinFieldElem}, s, os::Int)
+function hilbert90_cyclic(A::MatElem{<:FieldElem}, s, os::Int)
   #apart from rand, this would also work over a number field
+  k= base_ring(A)
```
```suggestion
  k = base_ring(A)
```

fingolfin commented on experimental/GModule/GModule.jl:
```diff
@@ -1453,6 +1566,7 @@ end
 
 function Oscar.simplify(C::GModule{<:Any, <:AbstractAlgebra.FPModule{ZZRingElem}})
  f = invariant_forms(C)[1]
+ global last_f = f
```
```suggestion
```

fingolfin commented on experimental/GaloisGrp/src/Subfields.jl:
```diff
@@ -109,6 +110,7 @@ function Base.:*(A::SubfieldLatticeElem, B::SubfieldLatticeElem)
   cs = B.b
   ds = [intersect(b, c) for b = bs for c = cs]
   ds = [x for x = ds if length(x) > 0]
+  sort!(ds)
```
Is this necessary, or for debugging, or what? (same for the other `sort!` calls you added)

fingolfin committed:
Apply suggestions from code review

fingolfin commented on experimental/GModule/GModule.jl:
```diff
@@ -112,7 +112,12 @@ case that `S` is a number field and `R` is the ring of integers in `S`,
 for example `R = ZZ` and `S = QQ`)
```
Those "examples" seem more like a TODO, perhaps you can also fill them in? (In a follow-up PR, though perhaps at least add "TODO" in this one so it won't be forgotten again.

The docstring above also claims 
> The syntax `M ⊗ phi` is supported.

which doesn't seem to be actually the case, so perhaps another TODO?

fingolfin reviewed (approved):
This seems to be passing tests now. For me it would be OK to merge this now as long as my remaining comments here are not forgotten but addressed in a follow-up PR.

fingolfin commented:
Thanks for merging, @fieker 

@mjrodgers as mentioned, please take care of the comments here in a follow-up PR :-)

benlorenz commented:
@fingolfin Do we really need this for the first 1.0 release, is there anything in the book that relies on this?

_Edit:_ was needed for the book and is backported now.

mjrodgers commented:
I will address them, currently working on the follow-up (and some G-Set documentation)
